### PR TITLE
Add model selection guide

### PR DESCRIPTION
**Pull request by smackesey**
This is a draft of the Model Selection guide. Parts of it are fairly polished, but there are two (short) unfinished sections at the end. I wouldn't spend your time pointing out typos or other nits yet, as I'll probably be refining it significantly. I'm opening the PR now though so that people can see where this is and discuss some broader questions:

First, this is different from other Dagster guides in that it requires a bit more explication of non-Dagster-specific stuff. Model selection is a somewhat tricky ML concept that is often discussed with confusing terminology. I thought it was necessary to explain it and try to be crystal clear in the mapping of ML concepts (e.g. model architecture) to programming constructs. However, as a result the beginning is a bit heavy on ML before getting to Dagster. This differs in tone from the rest of the guides (though it is similar to what you find in ML library docs). Personally I prefer this style, but I'm soliciting feedback here on whether I should cut out or adjust some of the ML explication.

Second, directions I want to take this guide (and potentially other ML guides) in:

- To appeal to the ML Persona, we really want to stress the unique benefits of Dagster over all the other ML pipeline/framework tools out there. I think I need to come up with a basic "pitch" to the ML Persona here, and bake it into this (and any other) ML guides, rather than just showing "Here's how to implement ML pattern X in Dagster". That's certainly useful, but we can do better if we really want to attract ML users.
- One specific benefit of Dagster is the fine-grained control over intermediates. This is particularly powerful when we have dagstermill ops that can generate notebook records/visualizations of those intermediates. Notebooks and probably the asset catalog should be used heavily in ML-focused guides.

Applying the above to the Model Selection guide, I plan to:

- (1) generally adjust text to advertise Dagster's unique benefits
- (2) replace the analysis ops with dagstermill implementations that include algorithm-specific visualizations for each trained model
- (3) generate a summary notebook at the end that pulls plots from all the intermediate models, giving you an extremely rich record of your analysis (scikit-learn gives you just the scores)

But posting this now to solicit any other feedback/ideas!

### CODE REVIEW

**Comment by sryza** [comment]
I think this is failing the docs build in some way that's preventing Vercel from deploying it (which would make it easier to review). Mind taking a look at that?

**Comment by smackesey** [comment]
It's now passing all the CI tests but Vercel still says "Canceled by Ignored Build Step". I looked a little at the Vercel docs and this means that there is a script configured to run on each new commit to determine if it should be deployed, and for some reason that script is saying the commits on this branch should *not* be deployed. I don't think I have access to the Vercel config (I don't see any Vercel config file in the repo-- but I'm not familiar with Vercel generally), so I can't determine why this is.

I'm pretty sure Vercel did *not* show the "Canceled by Ignored Build Step" on the first commit in this PR. So I wonder if it refusing to deploy subsequent commits has something to do with a lack of content changes between commits? My fixes were tweaks to sklearn dependency version (I originally had 1.0, which were causing python 3.6 tests to fail). If that's the case, we should tweak the Vercel config so that it keeps a link in the GH thread to the last commit that actually deployed for the branch, rather than a dead link to a deployment that was canceled.

**Comment by yuhan** [comment]
Re: Vercel skipping builds, I believe it was because those commits didn't update anything in the docs/ folder. We've configured the Vercel Git integration to skip building the docs site when `git diff HEAD^ HEAD --quiet docs/` exits code 0 (i.e. nothing changed in the docs/ for a given commit). 

As for the sklearn dependency version fix, the buildkite tests are independent from the Vercel builds, so the commit should be able to fix the python 3.6 tests and at the same time doesn't necessarily need to trigger a new build for docs site (the downgrade doesn't seem to have any impact on the content or infra for the docs site, so the previews shouldn't be different from previous successful builds).

**Comment by smackesey** [comment]
Thanks Yuhan, that all makes sense.

If I understand correctly, the issue is that the Vercel preview points to a branch, and the `sean/model-selection-guide` branch points to a commit that Vercel skipped deployment for, because docs content did not differ from a previous commit on the branch.

Should be able to fix this by just pushing a commit with a docs content update, but for the future, it seems like this is a bug/oversight in Vercel configuration. Seems to me the more useful behavior would be for the Vercel branch link to point to the last commit on the branch that was actually deployed, instead of being updated to point to nothing.

File: docs/content/guides/dagster/model_selection.mdx [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  target
+                   5.1               3.5                1.4               0.2       0
+                   4.9               3.0                1.4               0.2       0
+                   4.7               3.2                1.3               0.2       0
+                   4.6               3.1                1.5               0.2       0
+                   5.0               3.6                1.4               0.2       0
+
+The classification task is to predict the `target` using the physical measurements. Because this is a [supervised learning problem](https://machinelearningmastery.com/supervised-and-unsupervised-machine-learning-algorithms/), we have a clear model performance metric: the fraction of records correctly classified by the model. We'll call this number (ranging from 0-1) the _classification accuracy_.
```
Comment (sryza) [inline docs/content/guides/dagster/model_selection.mdx:31]: "we have a clear model performance metric"

Many would argue that accuracy is often the wrong metric for assessing the performance of classifiers.  E.g. https://stats.stackexchange.com/questions/312780/why-is-accuracy-not-the-best-measure-for-assessing-classification-models.

I'd remove the word "clear" and replace with something like "One performance metric we can use is accuracy"

File: docs/content/guides/dagster/model_selection.mdx [hunk 2]
```diff
@@ -0,0 +1,231 @@
+---
+title: Model Selection | Dagster
+description: This guide describes how to use Dagster for exploratory model selection for machine learning.
+---
+
+# Model Selection
+
+Every machine learning problem begins with exploration. Before training a final model or building an online learning system, there are many decisions to make. One of the most important decisions is: _what learning algorithm should be used, and how should it be configured_? This is known as _model selection._
+
+There are many learning algorithms for common tasks (e.g. classification, regression). Each algorithm defines a set of configurable hyperparameters, and each hyperparameter may range over a potentially infinite set of values. We can call a learning algorithm together with a fixed set of hyperparameter bindings a _model architecture_. Model selection then involves choosing a single model architecture out of an infinite set of possibilities.
```
Comment (sryza) [inline docs/content/guides/dagster/model_selection.mdx:10]: This might be a pedantic distinction, but a couple concerns with framing this as model architecture selection:
* I think many people understand the term "model architecture" to mean a learning algorithm _without_ the hyperparameters yet specified. E.g. one might say "we used a two-level CNN architecture, and we set the learning rate to 0.9".
* I suspect that, in many cases, the goal is model selection, not architecture/hyperparameter selection.  E.g. every week you want to train an ML model.  When you do that, you consider a bunch of architectures/hyperparameters.  But the next week, you don't just stick with the architecture/hyperparameters that worked best in the previous week - you again try training a bunch of models with a bunch of architectures/hyperparameters and again use the best one.

File: docs/content/guides/dagster/model_selection.mdx [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+## Dagster solution
+
+We're going to implement our solution using a combination of Dagster and popular machine learning library [Scikit Learn](https://scikit-learn.org/stable/index.html) (`sklearn`). We'll implement the orchestration aspects of our solution in Dagster, and outsource the fitting of individual models to `sklearn`. (Note: `sklearn` has its own powerful and concise API for model selection, as well as a limited pipeline API. Both are useful, but Dagster gives us much greater control over and observability of our computation graph.)
```
Comment (sryza) [inline docs/content/guides/dagster/model_selection.mdx:37]: Maybe worth mentioning that Dagster allows parallelizing / distributing the computation.

File: docs/content/guides/dagster/model_selection.mdx [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+height={1848}
+/>
+
+And here is the job definition:
+
+```python file=/guides/dagster/model_selection/job.py startafter=job_start endbefore=job_end
+@job
+def model_selection():
+    training_set, test_set = load_iris_data()
+    lr_hparam_sets, rf_hparam_sets = generate_candidate_model_architectures()
```
Comment (sryza) [inline docs/content/guides/dagster/model_selection.mdx:54]: Nitpick: preference for spelling out "hyperparameter" instead of abbreviating to "hparam"

Comment (sryza) [reply docs/content/guides/dagster/model_selection.mdx:54]: What would you think about having the function with the `DynamicOut` yield sklearn model objects instead of  hyperparameter combinations? That would make it easy to add other kinds of models without needing to change the graph.  And would also make the graph simpler and avoid code duplication between the `score_x_architecture` functions?

File: docs/content/guides/dagster/model_selection.mdx [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+## Dagster solution
+
+We're going to implement our solution using a combination of Dagster and popular machine learning library [Scikit Learn](https://scikit-learn.org/stable/index.html) (`sklearn`). We'll implement the orchestration aspects of our solution in Dagster, and outsource the fitting of individual models to `sklearn`. (Note: `sklearn` has its own powerful and concise API for model selection, as well as a limited pipeline API. Both are useful, but Dagster gives us much greater control over and observability of our computation graph.)
+
+Let's start with an overview of the finished Dagster job we will build. Here is what the completed job looks like in [Dagit](/concepts/dagit/dagit):
+
+<Image
```
Comment (sryza) [inline docs/content/guides/dagster/model_selection.mdx:41]: Including this visual representation of the graph is a nice touch - I think will be very helpful.

File: python_modules/dagster/setup.py [hunk 6]
```diff
@@ -108,6 +108,7 @@ def get_version() -> str:
                 "pytest-xdist==2.1.0",
                 "pytest==6.1.1",
                 "responses==0.10.*",
+                "scikit-learn<1.0.0",  # scikit-learn 1.0 requires python 3.7
```
Comment (sryza) [inline python_modules/dagster/setup.py:111]: Including the scikit-learn dependency here would pull in scikit-learn for everyone who depends on Dagster. If you put it in the setup.py under docs_snippets, we'd avoid that problem (although I think it's already there).

File: docs/content/guides/dagster/model_selection.mdx (hunk 2 above)
Comment (smackesey) [reply docs/content/guides/dagster/model_selection.mdx:10]: I don't think these points are pedantic-- I think it's important to be very precise when talking about machine learning. IMO there's enormous confusion among people because not enough authors have been pedantic enough! 

This is framed as "model architecture selection" bc this is the most accurate description I came up with of what we are actually doing. Here's why:

As you correctly point out, our ultimate goal is **model selection**. However, the naive approach to this of "training a bunch of models with a bunch of architectures/hyperparameters and [using] the best one" is biased because it overfits the training data. You can't get away from this problem, and so you shouldn't select model with this straightforward method.

Instead, *the rigorous way to do model selection is via model architecture selection.* You train a bunch of models, evaluate the performance of these models as a means of selecting the best *model architecture*, and then train a single final model (using the optimal architecture) on your entire training set (none of the models used to evaluate candidate architectures are trained on the whole training set). That final model is your answer to the model selection problem.

Also ,I'm open to using a different term than "model architecture", and I agree that it is somewhat ambiguous-- unfortunately there does not seem to be a better available term of art for this important concept: a fully-configured (i.e. all hyperparameters specified) learning algorithm. Part of the ambiguity arises from the fact that  hyperparameters often (e.g. number of hidden units in a neural network) map directly to "spatial", i.e. architectural, characteristics of the algorithm (unlike learning rate). In any case, this very ambiguity is part of the reason the background is as long as it is. With no unambiguous dominant lexicon, I have to set up clear meanings for these terms within the article scope.

File: docs/content/guides/dagster/model_selection.mdx (hunk 4 above)
Comment (smackesey) [reply docs/content/guides/dagster/model_selection.mdx:54]: I experimented with having `DynamicOut` yield sklearn model objects but decided against this. The reason is that it makes the particular set of hyperparameters you are optimizing less clear. In this example, there are more hyperparameters for both kinds of models than just the ones we are searching over (they're just set to defaults in the sklearn objects). If we directly yield model objects, then we later have to re-extract the particular hps we are varying to state the result of optimization. It is less complex IMO to simply keep them out as a separate entity. It's actually these little bundles of HP subsets (not the model objects) that are the domain we are optimizing over.

As for the code duplication between the `score_x_architecture` functions, that will make more sense when these two ops are replaced with notebook ops that include algorithm-specific visualizations.

**Comment by yuhan** [comment]
Re: Vercel preview
> Should be able to fix this by just pushing a commit with a docs content update, but for the future, it seems like this is a bug/oversight in Vercel configuration. Seems to me the more useful behavior would be for the Vercel branch link to point to the last commit on the branch that was actually deployed, instead of being updated to point to nothing.

I think the link on this message is up-to-date with the branch (the last commit with docs changes), i.e. the preview does pick up [this commit](https://github.com/dagster-io/dagster/pull/5707/commits/8cf6636ac495a4820c4ee78aa4f19501c53b32ab). 
image

For more context, we configured it this way, as opposed to always the last commit, to avoid having too many concurrent builds which were usually necessary and caused long queuing in the Vercel deployments.

**Comment by cmpadden** [comment]
Closing as I'd suspect this is no longer relevant; feel free to re-open if you'd like.
